pc-parse/pcpatr/grammar.c

```diff
@@ -1973,6 +2060,180 @@ listp->pNext      = pData->pGrammar->pRuleTable;
 pData->pGrammar->pRuleTable = listp;
 }
 
+/*****************************************************************************
+ * NAME
+ *    has_double_optional_constraint
+ * ARGUMENTS
+ *    pDag    - rule constriants
+ *    pszName - name of optional non-terminal
+ * DESCRIPTION
+ *    Does pDag have a constraint with pszName that mentions another optional
+ *    non-terminal?
+ * RETURN VALUE
+ *    int
+ */
+static int has_double_optional_constraint(pDag, pszName, pNonterminals)
+PATRFeature* pDag;
+char* pszName;
+PATRNonterminal* pNonterminals;
+{
+PATRComplexFeature* flist;
+PATRComplexFeature* flist2;
+PATRNonterminal* nterm;
+if (pDag == NULL)
+{
+	return FALSE;
+}
+if (pDag->pFirstFeat || pDag->pSecondFeat)
+{
+	return has_double_optional_constraint(pDag->pFirstFeat,
+			pszName, pNonterminals) ||
+		has_double_optional_constraint(pDag->pSecondFeat,
+			pszName, pNonterminals);
+}
+assert(pDag->eType == PATR_COMPLEX);
```

Comment:
Does the assert crash the program in a Release build?  If not, then having an if statement following the assert with the same test seems possibly a bit safer.